File: http/src/main/java/wang/leal/ahel/http/api/observer/ApiObserver.java
package wang.leal.ahel.http.api.observer;

import wang.leal.ahel.http.api.Api;
import wang.leal.ahel.http.api.exception.ApiException;
import wang.leal.ahel.http.cancel.DisposeCancelable;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public abstract class ApiObserver<T> implements Observer<T> {
    private Object tag;

    public ApiObserver(){
        this.tag = "ApiObserver";
    }

    public ApiObserver(Object tag){
        this.tag = tag;
    }

    @Override
    public final void onSubscribe(Disposable d) {
        Api.cancel().add(tag,new DisposeCancelable(d));
    }

    @Override
    public final void onNext(T data) {
        try {
            onSuccess(data);
        }catch (Exception e){
            e.printStackTrace();
        }
    }

    @Override
    public final void onError(Throwable e) {
        try {
            if (e instanceof ApiException){
                ApiException apiError = (ApiException) e;
                onApiError(apiError.getCode(),apiError.getMessage(),apiError.getBody());
            }else {
                onFailure(e);
            }
        }catch (Exception failureError){
            failureError.printStackTrace();
        }finally {
            try {
                onFinal();
            }catch (Exception finalError){
                finalError.printStackTrace();
            }
        }
    }

    @Override
    public final void onComplete() {
        onFinal();
    }

    //Http响应成功并且服务端code返回成功
    protected abstract void onSuccess(T data);

    //Http响应成功但是服务端code报告异常
    protected abstract void onApiError(int errNo,String errMsg,String data);

    //Http或者解析等等返回的各种错误
    protected abstract void onFailure(Throwable e);

    //无论成功或者失败，最终都会进入这里
    protected void onFinal(){

    }
}

File: http/src/main/java/wang/leal/ahel/http/api/ApiObserver.kt
package wang.leal.ahel.http.api

import io.reactivex.rxjava3.disposables.Disposable
import wang.leal.ahel.http.api.converter.ApiException

open class ApiObserver<T> {
    internal fun onSubscribe(disposable: Disposable) {
        onStart(disposable)
    }
    internal fun onNext(data: T) {
        try {
            onSuccess(data)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    internal fun onError(e: Throwable) {
        try {
            e.printStackTrace()
            if (e is ApiException) {
                onApiError(e.code(), e.message(), e.data())
            } else {
                onFailure(e)
            }
        } catch (failureError: Exception) {
            failureError.printStackTrace()
        } finally {
            try {
                onFinal()
            } catch (finalError: Exception) {
                finalError.printStackTrace()
            }
        }
    }

    internal fun onComplete() {
        try {
            onFinal()
        } catch (finalError: Exception) {
            finalError.printStackTrace()
        }
    }

    open fun onStart(disposable: Disposable){}

    //Http响应成功并且服务端code返回成功
    open fun onSuccess(data: T) {}

    //有响应但是业务异常
    open fun onApiError(errNo: Int, errMsg: String?, data: String?) {}

    //超时，解析等等返回的各种错误
    open fun onFailure(e: Throwable) {}

    //无论成功或者失败，最终都会进入这里
    open fun onFinal() {}
}